### ISSUE TITLE
Canonical way for labeling PRs based on code changes

### ISSUE THREAD

**Issue by kargakis** [kubernetes/test-infra#5332]
It seems that we have two different ways to add labels on PRs based on the code paths they touch.

This is supported:

1) in OWNERS files, though I can't find it documented anywhere, it seems to be used in the community repo via the [`owners-label`](https://github.com/kubernetes/test-infra/blob/31a2e95173bd87d91e011d0edfd844b869fa89e7/mungegithub/mungers/owner-label.go) munger - see [here](https://github.com/kubernetes/test-infra/blob/31a2e95173bd87d91e011d0edfd844b869fa89e7/prow/repoowners/repoowners.go#L54), [here](https://github.com/kubernetes/test-infra/blob/31a2e95173bd87d91e011d0edfd844b869fa89e7/mungegithub/submit-queue/deployment/community/configmap.yaml#L9), and an example [here](https://github.com/kubernetes/community/blob/d7c80174e370028a42636189168829ba9f1b0ac3/sig-apps/OWNERS#L6).
2) We also have the [`path-label`](https://github.com/kubernetes/test-infra/blob/31a2e95173bd87d91e011d0edfd844b869fa89e7/mungegithub/mungers/path_label.go) munger that is used elsewhere in kubernetes and we also use it in openshift.

Do we plan to converge the two mungers into a single prow plugin? Do we need to keep the support inside OWNERS?

cc @cblecker @cjwagner @stevekuznetsov @spxtr 

/area prow
/area mungegithub

**Comment by spiffxp** [comment]
I think I'd rather see us move to OWNERS rather than path-label

**Comment by cblecker** [comment]
I agree that we should consolidate, and my preference would be owners-label. Path-label is config that resides at the moment with the munger itself, but owners-label distributes the config into the managed repo itself.

**Comment by kargakis** [comment]
We have use cases for labeling code that touches paths that are independent of OWNERS files. For example, we label all PRs touching `/vendor` because we want to exclude them from merging every time we need to rebase on top of latest Kubernetes. Or we label all PRs touching API-related code because they need special attention for review. This kind of labeling cannot be expressed via OWNERS.

**Comment by kargakis** [comment]
Or it could but it would be ugly, especially in the case we want to track api changes.

**Comment by cjwagner** [comment]
I planned to consolidate these mungers while migrating mungers to Prow. Specifically I was going to migrate just the `owners-label` munger.

@kargakis Labeling PRs that touch `/vendor` can be achieved by adding a `/vendor/OWNERS` file. Labeling the API-related code changes would be a lot uglier as you mentioned because the regular expressions used to match those changes currently only apply to some files in the directories. We could augment the OWNERS file label section to allow optional regexps so that labels are only applied to files in the dir or subdir that match the regexp. Example:
```
approvers:
- cjwagner
labels:
- kind/api-change,/types.go$
```

I think it is preferable to have the labeling configuration located in the repo with the source code it is associated with instead of managing a monolithic config.

**Comment by spiffxp** [comment]
@cjwagner maybe this is out of scope for this issue, but WDYT about the idea of (optionally?) using OWNERS at repo root to auto-label issues against the repo?

**Comment by cjwagner** [comment]
@spiffxp I think we should keep the list of auto-labels for PRs separate from those for issues so that they can differ. That change would also break the scope of OWNERS files; they are supposed to only apply to changes that modify their directory or subdirectories. Issues don't make any changes at all.
If we want to automatically label issues based on the repo they are in that could be achieved with a simple new plugin that has config mapping repos to labels.

An alternative solution available now is adding `/label some-label` to the issue template for the repo. This isn't a perfect solution because users could circumvent the labeling by deleting that line from the template before creating the issue, but it works now.

**Comment by kargakis** [comment]
> I think it is preferable to have the labeling configuration located in the repo with the source code it is associated with instead of managing a monolithic config.

Agreed. There is already a [`labels.yaml`](https://github.com/kubernetes/kubernetes/blob/master/labels.yaml) that holds all labels inside the repo and their color. Is it a more appropriate place to set the paths they should be applied to?

**Comment by kargakis** [comment]
Another thing is that we have a bunch of projects that do not use OWNERS files today (actually all projects apart from our core repo). This means that we will need to add OWNERS files in all of those repos in order to automate PR labeling. Checking in a single `labels.yaml` at the root of the repo sounds more intuitive and doesn't require inventing machinery for updating all OWNERS file in case of label additions/updates/removals.

**Comment by kargakis** [comment]
@kubernetes/sig-contributor-experience-proposals

**Comment by cjwagner** [comment]
@kargakis I think `labels.yaml` is no longer being used in favor of a label-sync tool that runs on all repos in the org so that labels are standardized across the entire org.
I agree that we shouldn't have to have OWNERS files in a bunch of directories to specify labels for paths that can be identified with a regex. I think it might be confusing to require a new `labels.yaml` file that is separate from OWNERS files when we already have one form of path based labeling implemented in OWNERS files.
We could instead allow a root level OWNERS_REGEX file that can be used to specify approvers, reviewers, and labels by regex instead of by directory via regular OWNERS files.

**Comment by kargakis** [comment]
> @kargakis I think labels.yaml is no longer being used in favor of a label-sync tool that runs on all repos in the org so that labels are standardized across the entire org.

It seems that the `label-sync` tool is using that file.

**Comment by cjwagner** [comment]
That tool uses [this](https://github.com/kubernetes/test-infra/blob/master/label_sync/labels.yaml) `labels.yaml`, not the `labels.yaml` that is stored at the repo root of many k8s repositories. The later was used by the `check-labels` munger.
The label-sync tool avoids using repo specific `labels.yaml` files in order to standardize labels across repos.

**Comment by kargakis** [comment]
Another thing is we need a way to avoid readding labels in certain cases like when the label is removed. For example, in the case where we label all api changes with a [`needs-api-review`](https://github.com/openshift/release/blob/7f803f9c126e2a55250d3c8b417e1435a7248722/cluster/ci/config/submit-queue/submit_queue.yaml#L298) label, we have been asked for a way to remove the label once the review is done. The more I talk about it, the more I believe this ought to be its own external plugin but I am wondering if others ever wanted such a thing (remove a label added by the path-label munger at a later point in time).

**Comment by kargakis** [comment]
Fixed by https://github.com/kubernetes/test-infra/pull/6521
